core/src/main/kotlin/com/justai/jaicf/api/routing/BotRoutingEngine.kt

```diff
@@ -0,0 +1,151 @@
+package com.justai.jaicf.api.routing
+
+import com.justai.jaicf.BotEngine
+import com.justai.jaicf.api.BotRequest
+import com.justai.jaicf.builder.Scenario
+import com.justai.jaicf.context.BotContext
+import com.justai.jaicf.context.DialogContext
+import com.justai.jaicf.context.ExecutionContext
+import com.justai.jaicf.context.RequestContext
+import com.justai.jaicf.context.manager.BotContextManager
+import com.justai.jaicf.helpers.logging.WithLogger
+import com.justai.jaicf.reactions.Reactions
+import java.util.*
+
+/**
+ * This class provides a [BotEngine] implementation with BotRouting feature.
+ * BotRouting provides runtime functionality to change a [BotEngine] used to process client requests.
+ *
+ *
+ * Scenario usage example:
+ * ```
+ * private val main = Scenario {
+ *  fallback {
+ *      if (request.input == "sc1") routing.route("sc1")
+ *      else if (request.input == "sc2") routing.route("sc2")
+ *      else reactions.say("Hello from MAIN bot!")
+ *  }
+ * }
+ *
+ * private val sc1 = Scenario {
+ *  fallback {
+ *      reactions.say("Hello from SC1")
+ *  }
+ * }
+ *
+ * private val sc2 = Scenario {
+ *  fallback {
+ *      reactions.say("Hello from SC1")
+ *  }
+ * }
+ * private val router = BotRoutingEngine(
+ *  main = BotEngine(main),
+ *  routables = mapOf("sc1" to BotEngine(sc1), "sc2" to BotEngine(sc2))
+ *  )
+ * ```
+ *
+ * See more usage examples at examples/multilingual-bot.
+ *
+ * @param main main engine to process requests.
+ * @param routables a map of engines with names to perform route from and to.
+ *
+ * @see BotRoutingApi scenario runtime api for routing.
+ * */
+class BotRoutingEngine(
+    private val main: BotEngine,
+    routables: Map<String, BotEngine>,
+) : BotEngine(MOCK_SCENARIO), WithLogger {
```

Comment:
Looks like `BotRoutingEngine` is only a decorator for multiple BotEngines.
It's not clear for me why we need `BotEngine` implementation here, not `BotApi`?

Reply:
@morfeusys 
1. BotAPI has no knowledge of contextManagers, but here we must use it.
2. BotAPI is impossible to use in BotTest (mostly because of contextManagers).

Reply:
1. Context manager is a part of `process()`, isn't it enough?
2. Maybe it's a time to make it possible to use BotApi in tests?

Reply:
ContextManager has no part in process, only BotContext is part of process.
This can be used only if we make `BotContextManager` part of BotApi interface

Reply:
[BotApi.process()](https://github.com/just-ai/jaicf-kotlin/blob/master/core/src/main/kotlin/com/justai/jaicf/api/BotApi.kt#L33)

Reply:
@morfeusys I've edited my comment before you've seen it. But still BotContextManager is nullable, because it's channel-provided. And we cannot use it until it's nullable.

I'd suggest adding defaultContextManager to BotApi.process() if we must support BotApi.
```kotlin
interface BotApi {
    val defaultContextManager: BotContextManager
    fun process(...){}
``

Reply:
Each `BotEngine` has it's own `defaultContextManager` that can be retrieved. You can add additional context manager for `BotRoutingEngine` if you need it.